### PR TITLE
[MRG] TVM function, test and readme

### PR DESCRIPTION
**Pull request by amritaravishankar**
[MRG] Conversion of model to TVM

#### Does this change require updates to documentation?
Yes, this change requires an update to documentation as it provides additional functionality to ML-Model-CI

### CODE REVIEW

File: modelci/hub/tvm_testing.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import tvm
```
Comment (HuaizhengZhang) [inline modelci/hub/tvm_testing.py:1]: copy our lisences here
https://github.com/cap-ntu/ML-Model-CI/blob/6f326ed1314ec778e93f7cf8f537b54e822f18c4/modelci/cli/modelhub.py#L1

File: modelci/hub/tvm_testing.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        if not override:  # file exist yet override flag is not set
+            print('Use cached model')
+            return True
+    model = model.eval()
+    try:
+        input_data = torch.randn(input_shape)
+        traced_model = torch.jit.trace(model, input_data).eval()
+    
+        # image for testing 
+        img = Image.open(img_path).resize((224, 224))
```
Comment (HuaizhengZhang) [inline modelci/hub/tvm_testing.py:24]: we have an image pre-processing function and you can check our code (I forget where) to use it.

File: modelci/hub/tvm_testing.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        # img = tf.expand_dims(img, 0).numpy()
+       
+        # Import graph to relay
+        shape_list = [(input_name, img.shape)]
+        model, params = relay.frontend.from_pytorch(traced_model, shape_list)
+        
+        # Relay build
+        target = "llvm"
+        target_host = "llvm"
+        ctx = tvm.cpu(0)    # TODO onlt provides CPU as of now, future work: GPU
```
Comment (HuaizhengZhang) [inline modelci/hub/tvm_testing.py:49]: so the device type should be the function's input args

**Review by HuaizhengZhang — COMMENTED** [review]
- [ ] move the function to /test directory
- [ ] add a readme here
- [ ] complete the to_tvm.py under the /converter/ directorty
- [ ] re-request a review and replace WIP with MRG

**Comment by HuaizhengZhang** [comment]
Read files in /test and hub/converter to follow our coding and organization style.

File: modelci/hub/tvm_testing.py (hunk 2 above)
Comment (univerone) [reply modelci/hub/tvm_testing.py]: Maybe you could have a reference here for the image processing
https://github.com/cap-ntu/ML-Model-CI/blob/e3713239de8404958d72fa4c0a03c1ca601ca0e4/modelci/experimental/finetuner/resnet_fine_tune_example.py#L39

File: modelci/hub/converter/to_tvm.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+            with tvm.transform.PassContext(opt_level=opt_level):
+                lib = relay.build(model, target=target, target_host=target_host, params=params)
+
+            print('TVM format converted successfully')
+            return True
+
+        except:
+            #TODO catch different types of error
+            print("This model is not supported as TVM format")
```
Comment (univerone) [inline modelci/hub/converter/to_tvm.py:99]: Maybe raise  `NotImplementedError`  is better here

File: modelci/hub/converter/to_tvm.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            if not override:  # file exists yet override flag is not set
+                print('Use cached model')
+                return True
+        model.eval()
+        try:
+            input_data = torch.randn(input_shape)
+            traced_model = torch.jit.trace(model, input_data).eval()
+            
+            # image for testing 
+            img = Image.open(img_path).resize((224, 224))
```
Comment (HuaizhengZhang) [inline modelci/hub/converter/to_tvm.py:56]: We do not only support image models.

Please check the other converters and make this part more general

**Review by HuaizhengZhang — COMMENTED** [review]
I did not find the readme. Please also upload to the convert/ directory

File: modelci/hub/converter/to_tvm.py (hunk 5 above)
Comment (YuanmingLeee) [reply modelci/hub/converter/to_tvm.py:56]: The general idea is to generate fake input from the `IOShape` provided to you.

**Review by YuanmingLeee — COMMENTED** [review]
1. Fix your coding style followed by Codacy hints.
2. Merge from the master branch first, there is some update to help you pass the Unitest CI. @amritaravishankar

**Comment by YuanmingLeee** [comment]
image
@amritaravishankar

**Comment by univerone** [comment]
@amritaravishankar The test environment of Github action supports `import tvm` currently,  you can merge the latest code to your branch to check if the test will pass

**Comment by amritaravishankar** [comment]
> @amritaravishankar The test environment of Github action supports `import tvm` currently, you can merge the latest code to your branch to check if the test will pass

File: modelci/hub/converter/to_tvm.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                ctx = tvm.cpu(0)    
+                    
+            # TODO only provides CPU as of now, future work: GPU
+            # else if processing=='gpu':
+            #     ctx=tvm.gpu(0)
+
+            with tvm.transform.PassContext(opt_level=opt_level):
+                lib = relay.build(model, target=target, target_host=target_host, params=params)
+
+            print('TVM format converted successfully')
```
Comment (univerone) [inline modelci/hub/converter/to_tvm.py:89]: It's better to use logger instead of print

File: modelci/hub/converter/to_tvm.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            return True
+
+         except Exception as e:
```
Comment (univerone) [inline modelci/hub/converter/to_tvm.py:92]: Mind the indent of this line

File: modelci/hub/converter/to_tvm.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            # Facing an import error while importing modelci.types.bo
+            #  shape_list = [IOShape(shape=img.shape, dtype=self.dtype, name=self.input_name)]
+            model, params = relay.frontend.from_pytorch(traced_model, shape_list)
+                
+            # Relay build
+            target = "llvm"
+            target_host = "llvm"
+
+            if processing=='cpu':
+                ctx = tvm.cpu(0)
```
Comment (univerone) [inline modelci/hub/converter/to_tvm.py:80]: The ctx variable is not used

File: modelci/hub/converter/to_tvm.py (hunk 6 above)
Comment (amritaravishankar) [reply modelci/hub/converter/to_tvm.py:89]: I'm getting the following import error when I try to import Logger. Hence I resorted to use print.

File "../modelci/utils/misc.py", line 21, in <module>
    from typing import _GenericAlias, _SpecialForm, Any  # noqa
ImportError: cannot import name '_GenericAlias'

**Comment by univerone** [comment]
@amritaravishankar Now the tvm  import error has been fixed, you can go ahead now.

File: tests/test_tvm_conversion.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        cls.img_path = download_testdata(cls.img_url, "cat.png", module="data")
+        cls.input_name = 'input_0'
+        cls.input_shape = [1, 3, 224, 224]
+        cls.opt_level = 3
+        cls.test_path = Path("test_tvm")
+        cls.processing='cpu'
+        cls.override = False
+        cls.dtype = "float32"
+
+    def test_to_tvm1(self):
```
Comment (HuaizhengZhang) [inline tests/test_tvm_conversion.py:48]: The name is not good.

Better use test_to_tvm_from_cv_models()

File: modelci/hub/converter/to_tvm.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import torch.jit
+from torchvision import transforms
+
+from tvm import relay
+
+from pathlib import Path
+from PIL import Image
+
+import sys
+sys.path.append('../')
```
Comment (huangyz0918) [inline modelci/hub/converter/to_tvm.py:29]: You do not need to declare the system path here if you have added the Python path to the project workspace.

File: modelci/hub/converter/to_tvm.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    supported_framework = ["pytorch"]
+
+    @staticmethod
+    def from_pytorch(model: torch.nn.Module, save_path: Path, img_path, input_shape, input_name, processing, opt_level, override: bool = False):
+        """Convert a PyTorch nn.Module into TorchScript.
+        """
+        if save_path.with_suffix('.zip').exists():
+            if not override:  # file exists yet override flag is not set
+                print('Use cached model')
```
Comment (huangyz0918) [inline modelci/hub/converter/to_tvm.py:46]: We have a logger class, please use that.

**Review by huangyz0918 — CHANGES_REQUESTED** [review]
Some places still need to be improved. You can fix after you make TVM converter work on other types of network expect image models.

**Pull request closed without merging**